pennylane/_queuing_context.py

```diff
@@ -0,0 +1,95 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the :class:`QueuingContext` abstract base class.
+"""
+import abc
+
+
+class QueuingContext(abc.ABC):
+    """Abstract base class for classes that exposes a queue for Operations.
+
+    In PennyLane, quantum gates are not attached to some kind of circuit object
+    by the user. Here, the *construction* of a quantum gate is used to realize this
+    behaviour. As the constructor does not know to which QNode (PennyLane's circuit
+    object) it should attach the quantum gate we need to use a global variable that
+    holds that information. The ``QueuingContext`` class realizes this by providing
+    access to the current QNode. Furthermore, it provides the flexibility to have
+    multiple objects record the creation of quantum gates.
```

Comment:
This part is a great description of the problem that is being solved here! One thing I'm wondering if this would be the place where it should go in. Maybe it would be cool if we created a `Architecture` page or something similar in the docs, and this would be a perfect section there (then there could be a link included here).

Overall this would for sure be out of scope for this PR I believe, so for now maybe we could just keep this description here.

Reply:
Yes, but I don't know what else would go into that `Architecture` page 🤔 But that's surely for another PR